### DISCUSSION TITLE
Does the node features get duplicated for each edge?

### DISCUSSION THREAD

**Discussion by HongtaoYang** [pyg-team/pytorch_geometric#2356] [Q&A] [answered]
Suppose I have a graph of **N** nodes and **E** edges.  During the propagation process, I can see that the `x_j` and `x_i` are of shape `[E, feature_dim]`, does that mean the node features are duplicated based on the edges?

Does it also mean that the memory footprint is O(E) instead of O(N)?

**Comment by rusty1s** [comment, marked as answer]
That is correct. Some operators can be implemented more efficiently though, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html).

**Reply by HongtaoYang** [reply]
Thanks!
